src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -357,564 +398,511 @@ public override unsafe string GetString(byte[] bytes, int byteIndex, int byteCou
             {
-                // First unwind any fallback
-                if (ch == 0)
+                // Extract the next UCS-2 character and run it through our fallback mechanism.
+                // Also check that we didn't integer overflow while computing the destination size.
+
+                totalByteCount += fallbackBuffer.InternalGetFallbackByteCount(pinnedChars[0], this);
+                if (totalByteCount < 0)
                 {
-                    // No fallback, just get next char
-                    ch = *chars;
-                    chars++;
+                    goto Overflow;
                 }
 
-                // Check for fallback, this'll catch surrogate pairs too.
-                // no chars >= 0x80 are allowed.
-                if (ch > 0x7f)
+                pinnedChars = pinnedChars.Slice(1); // we consumed a single char
+
+                if (!pinnedChars.IsEmpty)
                 {
-                    if (fallbackBuffer == null)
+                    // If more input data remains, count the number of ASCII characters.
+
+                    indexOfFirstNonAsciiChar = (int)AsciiUtility.GetIndexOfFirstNonAsciiChar((char*)Unsafe.AsPointer(ref MemoryMarshal.GetReference(pinnedChars)), (uint)pinnedChars.Length);
+                    Debug.Assert(0 <= indexOfFirstNonAsciiChar && indexOfFirstNonAsciiChar <= pinnedChars.Length);
+
+                    // Check that we didn't integer overflow when computing the total byte count.
+
+                    totalByteCount += indexOfFirstNonAsciiChar;
+                    if (totalByteCount < 0)
                     {
-                        // Initialize the buffer
-                        if (encoder == null)
-                            fallbackBuffer = this.encoderFallback.CreateFallbackBuffer();
-                        else
-                            fallbackBuffer = encoder.FallbackBuffer;
-                        fallbackBuffer.InternalInitialize(charEnd - charCount, charEnd, encoder, false);
+                        goto Overflow;
                     }
 
-                    // Get Fallback
-                    charsForFallback = chars; // Avoid passing chars by reference to allow it to be enregistered
-                    fallbackBuffer.InternalFallback(ch, ref charsForFallback);
-                    chars = charsForFallback;
-                    continue;
+                    // Skip over all the ASCII data we just checked.
+
+                    pinnedChars = pinnedChars.Slice(indexOfFirstNonAsciiChar);
                 }
+            };
 
-                // We'll use this one
-                byteCount++;
-            }
+        Finished:
 
-            Debug.Assert(fallbackBuffer == null || fallbackBuffer.Remaining == 0,
-                "[ASCIIEncoding.GetByteCount]Expected Empty fallback buffer");
+            return totalByteCount;
 
-            return byteCount;
+        Overflow:
+
+            ThrowConversionOverflow();
```

Comment:
We do not use ThrowHelper is non-generic non-perf-critical code. The savings from using them are not worth the extra complexity.